File: themr.py
#!/bin/env python3
"""A program to set a global theme by replacing strings in config files."""

import re
import argparse
import yaml
import subprocess
from os import environ
from os.path import expanduser


debug = False


def main():
    """Entry point."""
    global debug

    parser = argparse.ArgumentParser(
        description="Set a theme in multiple programs by replacing strings in their config files."
    )

    parser.add_argument('theme_name', metavar='THEME', type=str, nargs='?', help='theme to use')
    parser.add_argument('-l', '--list-themes', action='store_true', help="list supported themes and exit")
    parser.add_argument('-c', '--list-configs', action='store_true', help='list supported configs and exit')
    parser.add_argument('-d', '--debug', action='store_true', help='print debug messages')

    args = parser.parse_args()

    debug = args.debug
    chosen_theme_name = args.theme_name

    config_folder_path = environ.get("XDG_CONFIG_HOME", expanduser("~/.config"))
    config_folder_path += "/themr"

    configs = load_configs(config_folder_path)
    themes, invalid_themes = load_themes(config_folder_path, [name for name in configs.keys()])

    if args.list_themes:
        list_themes(themes)

    if args.list_configs:
        list_configs(configs)

    if args.list_themes or args.list_configs:
        exit(0)

    if chosen_theme_name is None:
        print("No theme name given")
        exit(1)
    elif chosen_theme_name in invalid_themes:
        print(f"Theme {chosen_theme_name} exists but was ignored due to the errors above.")
        exit(1)

    if (chosen_theme := themes.get(chosen_theme_name)) is None:
        print(f'Theme "{chosen_theme_name}" does not exist')
        exit(1)

    set_theme(chosen_theme, configs)


def load_themes(folder_path, config_names):
    """Load and validate themes from file."""

    invalid_themes = list()

    try:
        with open(f'{folder_path}/themes.yaml') as themes_stream:
            themes = yaml.safe_load(themes_stream)  # TODO: use a class for themes
    except FileNotFoundError:
        print("themes.yaml was not found!")
        exit(-1)

    for theme_name, theme in dict(themes).items():
        valid = True

        if not theme.get("default"):
            for config_name in config_names:
                if config_name not in theme.keys():
                    print(f'Missing value for "{config_name}" in {theme_name} theme, while a default value is not given.')

                    valid = False

            if not valid:
                print(f'Ignored theme: {theme_name}.')
                themes.pop(theme_name)
                invalid_themes.append(theme_name)

    return themes, invalid_themes


def load_configs(folder_path):
    """Load and validate configs from file."""
    try:
        with open(expanduser(f'{folder_path}/configs.yaml')) as configs_stream:
            configs = yaml.safe_load(configs_stream)  # TODO: use a class for themes
    except FileNotFoundError:
        print("configs.yaml was not found!")
        exit(-1)

    required_keys = ['path', 'regex', 'pre', 'post']

    for config_name, config in dict(configs).items():
        # make name available like the rest of the values
        config["name"] = config_name

        valid = True

        for key in required_keys:
            if key not in config.keys():
                print(f'Missing value for "{key}" in {config["name"]} config.')

                valid = False

        if not valid:
            print(f'Ignoring {config_name} config.')
            configs.pop(config_name)

    return configs


def list_themes(themes):
    global debug

    print("Found themes:")
    for theme in themes:
        print("\t" + theme)

    if debug:
        print(themes)


def list_configs(configs):
    global debug

    print("Found configs:")
    for config in configs:
        print("\t" + config)

    if debug:
        print(configs)


def get_new_config_contents(theme, config):
    warnings = list()
    success = False

    config_name = config["name"]
    config_path = config["path"]
    regex = re.compile(config["regex"])

    # If we don't have a specific theme name, use the default.
    # This is done so if a theme has the same name in every config,
    # you don't have to repeat it each time.
    replace = theme.get(config_name, theme["default"])

    new_config_contents = list()
    try:
        with open(expanduser(config_path), 'r') as config_file:
            for line in config_file:
                if regex.match(line) and not success:
                    new_config_contents.append(config["pre"] + replace + config["post"] + '\n')
                    if debug:
                        print(f'found line:  {line}', end='')
                        print(f'replaced by: {config["pre"] + replace + config["post"]}')
                    success = True
                else:
                    new_config_contents.append(line)
    except FileNotFoundError:
        warnings.append(f'No file found at: "{config["path"]}')

    if not success:
        warnings.append(f'No line matched regex: "{config["regex"]}"')

    return new_config_contents, warnings


def write_new_config_file(config, contents):
    config_path = expanduser(config["path"])

    with open(config_path, 'w') as config_file:
        for line in contents:
            config_file.write(line)


def set_theme_for_config(theme, config):
    warnings = list()

    if debug:
        print(f'Working on {config["name"]}')

    new_config_contents, warn = get_new_config_contents(theme, config)

    if not warn:
        write_new_config_file(config, new_config_contents)
    else:
        warnings.extend(warn)

    try:
        cmd = config["cmd"]
        replace = theme.get(config["name"], theme["default"])

        if '%' in cmd:
            cmd = cmd.replace("%", replace)

        subprocess.Popen(cmd, shell=True, stdin=None, stdout=None)

        if debug:
            print(f'Executing post replace command: "{cmd}"')

    except KeyError:
        pass

    if warnings:
        print(f'Warnings from {config["name"]} config:')
        for warn in warnings:
            print('\t' + warn + '\n')

    return warnings


def set_theme(theme, configs):
    """Set the theme in configs."""
    global debug

    for config_name, config in configs.items():

        warnings = set_theme_for_config(theme, config)

        if warnings:
            print(f'Warnings from {config_name} config:')
            for warn in warnings:
                print('\t' + warn + '\n')


if __name__ == "__main__":
    main()

File: themr.go
package main

import (
	"errors"
	"fmt"
	"os"
	"os/user"
	"path/filepath"
	"strings"
	"sync"
	"runtime"

	"themr/config"

	"github.com/charmbracelet/log"
	"github.com/hellflame/argparse"
	"gopkg.in/yaml.v3"
)

var (
	logger log.Logger
	debug  *bool
)

const VERSION = "0.2.4"

func init() {
	logger = log.New(log.WithTimeFormat(""))
}

func main() {
	parser := argparse.NewParser("themr", "Set a theme in multiple programs by replacing strings in their config files.", &argparse.ParserConfig{DisableDefaultShowHelp: true})

	chosen_theme_name := parser.String("", "theme", &argparse.Option{Positional: true})
	list_configs_flag := parser.Flag("c", "list-configs", &argparse.Option{Help: "list supported configs"})
	list_themes_flag := parser.Flag("l", "list-themes", &argparse.Option{Help: "list supported themes"})
	print_version := parser.Flag("v", "version", &argparse.Option{Help: "print version"})
	debug = parser.Flag("d", "debug", &argparse.Option{Help: "print debug messages"})

	if e := parser.Parse(nil); e != nil {
		// check if empty error, happens with -h flag
		if msg := e.Error(); msg != "" {
			logger.Error(e.Error())
		}
		os.Exit(1)
	}

	if *debug {
		logger.SetLevel(log.DebugLevel)
	}

	config.SetLogger(logger)

	if *print_version {
		fmt.Println(os.Args[0], "v"+VERSION)
		os.Exit(0)
	}

	// get config path
	config_dir, err := os.UserConfigDir()
	config_dir += "/themr/"
	if err != nil {
		logger.Error("Could not determine User Config Directory. (is $HOME unset?)")
		os.Exit(1)
	}

	// Override config directory to use ~/.config on macOS
	if runtime.GOOS == "darwin" {
			config_dir = filepath.Join(os.Getenv("HOME"), ".config")
			config_dir += "/themr/"
	}

	if *debug {
		logger.Debug(config_dir)
	}
	edits, err := config.Load_configs(config_dir)
	if err != nil {
		logger.Error(err.Error())
		os.Exit(1)
	}
	// valid config types is the set resulting from
	// the union of config names and their types
	config_types := make(set)
	var str_configs string
	for _, configs := range edits {
		for _, config := range configs {
			str_configs += config.Name + "\n"
			config_types[config.Name] = member
			if config_type := config.Type; config_type != "" {
				config_types[config_type] = member
			}
		}
	}
	logger.Debug(fmt.Sprintf("loaded %d configs", len(edits)), "names", str_configs)

	// load themes
	themes, err := load_themes(config_dir, config_types)
	if err != nil {
		logger.Error(err.Error())
		os.Exit(1)
	}
	var str_themes string
	for _, theme := range themes {
		str_themes += theme["name"] + "\n"
	}
	logger.Debug(fmt.Sprintf("loaded %d themes", len(themes)), "names", str_themes)

	if *list_configs_flag {
		list_configs(edits)
		os.Exit(0)
	}

	if *list_themes_flag {
		list_themes(themes)
		os.Exit(0)
	}

	if *chosen_theme_name == "" {
		logger.Error("No theme name given")
		os.Exit(1)
	}

	var chosen_theme theme_info
	for _, theme := range themes {
		if theme["name"] == *chosen_theme_name {
			chosen_theme = theme
		}
	}

	if chosen_theme == nil {
		logger.Error("No such theme exists")
		os.Exit(1)
	}

	chosen_theme.set(edits)
}

func (t theme_info) set(edits config.Edits) {

	// only keep configs with appropriate types
	// var configs []config.Config
	for path, configs := range edits {
		for i, config := range configs {
			if !t.Map().contains_key(config.Type) {
				if len(edits[path]) == 1 { // if only one config left for this edit
					delete(edits, path)
				} else { // else remove only the 1 config
					edits[path] = append(edits[path][:i], edits[path][i+1:]...)
				}
			}
		}
	}

	if *debug {
		for _, configs := range edits {
			for _, config := range configs {
				t.set_for(config)
			}
		}
		return
	}

	var wg sync.WaitGroup
	for _, configs := range edits {
		wg.Add(1)
		go func(t theme_info, configs []config.Config) {
			for _, conf := range configs {
				defer wg.Done()
				t.set_for(conf)
			}
		}(t, configs)
	}
	wg.Wait()
}

func (t theme_info) set_for(config config.Config) {

	path := config.Path
	if strings.HasPrefix(path, "~") {
		usr, _ := user.Current()
		path = filepath.Join(usr.HomeDir, path[2:])
	}

	// use theme name for the type of config
	theme_name := t[config.Type]
	// unless it's overwitten by a theme specifying a theme_name for a config
	if name, exists := t[config.Name]; exists {
		theme_name = name
	}

	// warn if directory of config does not exist
	_, err := os.Stat(filepath.Dir(path))
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			logger.Warn("Path to config does not exist.\n\tMaybe you forgot to stow something?")
		} else {
			logger.Error("Stat() config dir: " + err.Error())
		}
	}

	file, err := os.ReadFile(path)
	// if the config tells use to create the file if it doesn't exist
	// we just place the "Replace" string in the "file"'s contents
	// and write those back
	if err != nil {
		if !config.Create {
			logger.Error("Can't read file: " + err.Error())
			return
		}
		file = []byte(config.Replace)
	}
	if !config.Regex.Match(file) {
		logger.Error("Configuration: Regex `" + config.Regex.String() + "` failed to match in file: " + config.Path)
		return
	}

	line := strings.ReplaceAll(config.Replace, "{}", theme_name)
	new_contents := config.Regex.ReplaceAll(file, []byte(line))

	// create path to config, incase iy does not exist
	// err = os.MkdirAll(filepath.Dir(path), os.FileMode(0700))
	// if err != nil {
		// logger.Error("Can't create directory path to config: " + err.Error())
	// }

	// write back the file :)
	err = os.WriteFile(path, new_contents, os.FileMode(0664))
	if err != nil {
		logger.Error("Can't save: " + err.Error())
	}

	err = config.RunCmd(theme_name, *debug)
	if err != nil {
		logger.Warn(fmt.Errorf("Command for "+config.Name+" failed: %w", err).Error())
	}
}

func list_themes(themes []theme_info) {
	fmt.Println("Found themes:")

	for _, theme := range themes {
		fmt.Println("\t" + theme["name"])
	}
}

func list_configs(edits config.Edits) {
	fmt.Println("Found configs:")

	for _, configs := range edits {
		for _, config := range configs {
			fmt.Println("\t" + config.Name)
		}
	}
}

func load_themes(config_dir string, config_types set) ([]theme_info, error) {
	theme_path := config_dir + "themes.yaml"
	themes := make(map[string]theme_info)

	file, err := os.ReadFile(theme_path)

	if err != nil {
		return nil, err
	}

	yaml.Unmarshal(file, &themes)

	var themes_list []theme_info

	for theme_name, theme := range themes {
		theme["name"] = theme_name

		// check if theme contains at least one config type
		if !theme.Map().contains_at_least_one_key(config_types) {
			return nil, fmt.Errorf("Theme must have at least one config type: '" + theme["name"] + "' does not!")
		}

		themes_list = append(themes_list, theme)
	}

	return themes_list, err
}
